docs/csharp/language-reference/operators/bitwise-and-shift-operators.md

```diff
@@ -0,0 +1,182 @@
+---
+title: "Bitwise and shift operators - C# Reference"
+description: "Learn about C# operators that perform bitwise logical or shift operations with the operands of the integral types."
+ms.date: 04/16/2019
+author: pkulikov
+f1_keywords: 
+  - "~_CSharpKeyword"
+  - "<<_CSharpKeyword"
+  - ">>_CSharpKeyword"
+  - "&_CSharpKeyword"
+  - "^_CSharpKeyword"
+  - "|_CSharpKeyword"
+helpviewer_keywords: 
+  - "bitwise logical operators [C#]"
+  - "shift operators [C#]"
+  - "tilde operator [C#]"
+  - "one's complement operator [C#]"
+  - "bitwise complement operator [C#]"
+  - "~ operator [C#]"
+  - "left shift operator [C#]"
+  - "<< operator [C#]"
+  - "right shift operator [C#]"
+  - ">> operator [C#]"
+  - "bitwise logical AND operator [C#]"
+  - "ampersand operator [C#]"
+  - "& operator [C#]"
+  - "bitwise logical exclusive OR operator [C#]"
+  - "bitwise logical XOR operator [C#]"
+  - "^ operator [C#]"
+  - "bitwise logical OR operator [C#]"
+  - "| operator [C#]"
+---
+# Bitwise and shift operators (C# Reference)
+
+The following operators perform bitwise or shift operations with the operands of the [integral types](../keywords/integral-types-table.md):
```

Comment:
The shift operators are only defined on `int`, `uint`, `long`, and `ulong`. They are not defined on other integral types (Note that I can't say why they aren't defined on `ushort`, or `short`, but it's what the [spec says](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#shift-operators).)

The same is true for the bitwise complement (`~`) operator. See [spec link](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#bitwise-complement-operator)

And also true for the [logical operators](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#logical-operators).

Yes, I had to check the spec to make sure on this one. I always enjoy those days.

Reply:
@BillWagner yes, they are defined for only those four, but another four types are also *supported*. That is, operands might be of `byte`, but they would be converted to `int` before the operation. The same takes place for the arithmetic operators. However, we don't stress that there. That's why, in the text, I don't say that operators are *defined*, but perform operations. However I silence the fact that less-than-32-bit operands are promoted to `int` in the process of such an operation. I assumed that knowledge  is for the spec. Do you think *this* overview should be more explicit about integer operand promotion? Then, I will think how to update the introduction.

Reply:
🤔 It might be more important for these operators than the arithmetic operators. My reasoning is that the result is the promoted type. 

```csharp
byte b = 0x88;
var shifted = b >> 3;
Console.WriteLine(shifted.GetType()); // "System.Int32"
```

In the above example, it doesn't matter, but it matter when shifting left. The result might be truncated if stored in its original type:

```csharp
    byte b = 0x88;
    var shifted = b << 3;
    Console.WriteLine(shifted.GetType());
    Console.WriteLine(shifted); // 1088
    byte b2 = (byte) shifted;
    Console.WriteLine(b2);  // 64
```

What are your thoughts?

Reply:
@BillWagner 

> It might be more important for these operators than the arithmetic operators.

> ...it matter when shifting left.

Agree, in the end, those bit operations work not with the numbers but rather with the sets of bits. "Exploding" a bit set in size is indeed quite important to be mentioned, especially, for the shift operations. 

I'll update the text and add an example for the left shift. Thanks for pointing out to the matter.